language/move-prover/src/spec_translator.rs

```diff
@@ -1426,28 +1431,110 @@ impl<'env> SpecTranslator<'env> {
 
     pub fn emit_on_update_global_invariant_assumes(&self, memory: QualifiedId<StructId>) {
-        let env = self.global_env();
         self.emit_global_invariants(
             true,
-            env.get_global_invariants_for_memory(memory)
-                .into_iter()
-                .filter(|id| {
-                    env.get_global_invariant(*id)
-                        .and_then(|inv| {
-                            env.is_property_true(&inv.properties, CONDITION_ON_UPDATE_PROP)
-                        })
-                        .unwrap_or(false)
-                })
-                .collect_vec(),
+            self.get_effective_global_invariants(memory, GlobalInvariantContext::AssumeOnUpdate),
         );
     }
 
-    pub fn emit_global_invariants_for_memory(&self, assume: bool, memory: QualifiedId<StructId>) {
-        let env = self.global_env();
-        self.emit_global_invariants(assume, env.get_global_invariants_for_memory(memory))
+    pub fn emit_on_access_global_invariant_assumes(&self, memory: QualifiedId<StructId>) {
+        self.emit_global_invariants(
+            true,
+            self.get_effective_global_invariants(memory, GlobalInvariantContext::AssumeOnAccess),
+        );
+    }
+
+    pub fn emit_on_update_global_invariant_asserts(&self, memory: QualifiedId<StructId>) {
+        self.emit_global_invariants(
+            false,
+            self.get_effective_global_invariants(memory, GlobalInvariantContext::Assert),
+        );
     }
 
-    fn emit_global_invariants(&self, assume: bool, invariants: Vec<GlobalId>) {
+    /// Returns the set of effective global invariants for a given memory in a given usage
+    /// context (on top-level function entry, on memory access, and before and after a memory
+    /// update). This uses properties on the invariants and program options to determine the
+    /// effective set.
+    fn get_effective_global_invariants(
+        &self,
+        mem: QualifiedId<StructId>,
+        context: GlobalInvariantContext,
+    ) -> Vec<GlobalId> {
+        // All invariants which refer to this memory.
+        let all_invariants = self.global_env().get_global_invariants_for_memory(mem);
+
+        // A predicate determining whether the invariant has the property `isolated`. Such
+        // invariants are not used as assumptions for other verification steps.
+        let is_isolated = |id: &GlobalId| {
+            self.global_env()
+                .get_global_invariant(*id)
+                .and_then(|inv| {
+                    self.global_env()
+                        .is_property_true(&inv.properties, CONDITION_ISOLATED)
+                })
+                .unwrap_or(false)
+        };
+        let is_connected = |id: &GlobalId| !is_isolated(id);
+
+        // A predicate which determines whether this invariant depends on any memory which
+        // is part of the verification problem, that is touches a memory which is declared by
+        // a module which is verified.
```

Comment:
Does "a module which is verified" refer to the modules passed as arguments on the command line to Move Prover?

Reply:
Yes